src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,249 @@
+  groups?: string[];
+}
+
+export type Permissions = Record<string, Principals>;
+
+export interface TransformedPermission {
+  type: string;
+  name: string;
+  permissions: string[];
+}
+
+const addToPrincipals = (principals?: Principals, users?: string[], groups?: string[]) => {
+  if (!principals) {
+    principals = {};
+  }
+  if (!!users) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deleteFromPrincipals = (principals?: Principals, users?: string[], groups?: string[]) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const checkPermission = (currentPrincipals: Principals | undefined, principals: Principals) => {
+  return (
+    (currentPrincipals?.users &&
+      principals?.users &&
+      checkPermissionForSinglePrincipalType(currentPrincipals.users, principals.users)) ||
+    (currentPrincipals?.groups &&
+      principals.groups &&
+      checkPermissionForSinglePrincipalType(currentPrincipals.groups, principals.groups))
+  );
+};
+
+const checkPermissionForSinglePrincipalType = (
+  currentPrincipalArray: string[],
+  principalArray: string[]
+) => {
+  return (
+    currentPrincipalArray &&
+    principalArray &&
+    (currentPrincipalArray.includes('*') ||
+      principalArray.some((item) => currentPrincipalArray.includes(item)))
+  );
+};
+
+export class ACL {
+  private permissions?: Permissions;
+  constructor(initialPermissions?: Permissions) {
+    this.permissions = initialPermissions || {};
+  }
+
+  // parse the permissions object to check whether the specific principal has the specific permission types or not
+  public hasPermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || permissionTypes.length === 0 || !this.permissions || !principals) {
+      return false;
+    }
+
+    const currentPermissions = this.permissions;
+    return permissionTypes.some((permissionType) =>
+      checkPermission(currentPermissions[permissionType], principals)
+    );
+  }
+
+  // permissions object build function, add principal with specific permission to the object
+  public addPermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || !principals) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
+      this.permissions[permissionType] = addToPrincipals(
```

Comment:
do we need to assign the result back if it's already updated in addToPrincipals?